airflow/providers/amazon/aws/operators/mysql_to_s3.py

```diff
@@ -0,0 +1,126 @@
+                (unless use_ssl is False), but SSL certificates will not be verified.
+        - ``path/to/cert/bundle.pem``: A filename of the CA cert bundle to uses.
+                You can specify this argument if you want to use a different
+                CA cert bundle than the one used by botocore.
+    :type verify: bool or str
+    :param pd_csv_kwargs: arguments to include in pd.to_csv (header, index, columns...)
+    :type pd_csv_kwargs: dict
+    :param index: whether to have the index or not in the dataframe
+    :type index: str
+    :param header: whether to include header or not into the S3 file
+    :type header: bool
+    """
+
+    template_fields = ('s3_key', 'query',)
+    template_ext = ('.sql',)
+
+    @apply_defaults
+    def __init__(
+            self,
+            query: str,
+            s3_bucket: str,
+            s3_key: str,
+            mysql_conn_id: str = 'mysql_default',
+            aws_conn_id: str = 'aws_default',
+            verify: Optional[Union[bool, str]] = None,
+            pd_csv_kwargs: Optional[dict] = None,
+            index: Optional[bool] = False,
+            header: Optional[bool] = False,
+            *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.query = query
+        self.s3_bucket = s3_bucket
+        self.s3_key = s3_key
+        self.mysql_conn_id = mysql_conn_id
+        self.aws_conn_id = aws_conn_id
+        self.verify = verify
+
+        if not pd_csv_kwargs:
+            self.pd_csv_kwargs = {}
+        else:
+            self.pd_csv_kwargs = pd_csv_kwargs
+        if "index" not in self.pd_csv_kwargs:
+            self.pd_csv_kwargs["index"] = index
+        if "header" not in self.pd_csv_kwargs:
+            self.pd_csv_kwargs["header"] = header
+
+    def _fix_int_dtypes(self, df):
+        """
+        Mutate DataFrame to set dtypes for int columns containing NaN values."
+        """
+        for col in df:
+            if "float" in df[col].dtype.name and df[col].hasnans:
+                # inspect values to determine if dtype of non-null values is int or float
+                notna_series = df[col].dropna().values
+                if np.isclose(notna_series, notna_series.astype(int)).all():
+                    # set to dtype that retains integers and supports NaNs
+                    df[col] = np.where(df[col].isnull(), None, df[col]).astype(pd.Int64Dtype)
+
+    def execute(self, context):
+        mysql_hook = MySqlHook(mysql_conn_id=self.mysql_conn_id)
+        s3_conn = S3Hook(aws_conn_id=self.aws_conn_id, verify=self.verify)
+        data_df = mysql_hook.get_pandas_df(self.query)
+        self.log.info("Data from MySQL obtained")
+
+        self._fix_int_dtypes(data_df)
+        with tempfile.NamedTemporaryFile(mode='r+', suffix='.csv') as tmp_csv:
+            tmp_csv.file.write(data_df.to_csv(**self.pd_csv_kwargs))
+            tmp_csv.file.seek(0)
```

Comment:
What happens if you pass a file to the current code (`tmp_csv.file.write(data_df.to_csv(**self.pd_csv_kwargs))`) through `pd_csv_kwargs`? 

**Suggestion 1:**

Can't we do it like this i.e. write to the file through `to_csv` ?
```suggestion
       with tempfile.NamedTemporaryFile(mode='r+', suffix='.csv') as tmp_csv:
           data_df.to_csv(tmp_csv.name, **self.pd_csv_kwargs)
```
And validate `pd_csv_kwargs` to not contain `path_or_buf` because you create the file.
```
        if "path_or_buf" in self.pd_csv_kwargs:
            raise
```

**Suggestion 2:**

Write to s3 through `load_string` and pass the result of `data_df.to_csv(**self.pd_csv_kwargs)`. This way you do not need to create a temp file here. The file will be created by s3 hook.

But you should also validate `pd_csv_kwargs` to not contain `path_or_buf`.

Reply:
Thank you very much for your suggestions. 
Regarding validating pd_csv_kwargs, I will do it.

Regarding the other suggestion, I already tried to use just to_csv and I failed. Also, at the beginning, I was using StringIO (equivalent to BytesIO that s3 hook uses) and I was suggested to use the way I am using right now. Everything can be found in this discussion:
https://github.com/apache/airflow/pull/6670#discussion_r361645289

Reply:
okay, but then **Solution 1** should work. You should pass the file name not the object. Can you try this? `data_df.to_csv(tmp_csv.name, **self.pd_csv_kwargs)`

Reply:
I made the changed and it worked. Thank you very much. Is there anything else I should to have your approval @feluelle ?